version.go

```diff
@@ -471,6 +472,79 @@ func (v Version) MarshalText() ([]byte, error) {
 	return []byte(v.String()), nil
 }
 
+// UnmarshalBinary implements the encoding.BinaryUnmarshaler interface.
+func (v *Version) UnmarshalBinary(data []byte) (err error) {
+	var nums [3]uint64
+	r := bytes.NewReader(data)
+	for i := range nums {
+		nums[i], err = binary.ReadUvarint(r)
+		if err != nil {
+			return ErrInvalidSemVer
+		}
+	}
+
+	var strings [2][]byte
+	var i int
+	// range doesn't advance i to 2
+	// (used to determine whether `break` happened)
+	for i = 0; i < len(strings); i++ {
+		length, err := binary.ReadUvarint(r)
+		if err != nil {
+			break
+		}
+		if length == 0 {
+			continue
+		}
+		if length > uint64(r.Len()) {
+			// Parsed length can't be larger than the remaining
+			// length of the data; without this check a maliciously crafted
+			// `data` can make us attempt to allocate a huge buffer
+			break
+		}
+		// We allocate here instead of returning a slice of data:
+		// data might be a slice of some enormous binary blob. It would be bad
+		// if our reference to the tiny prerelease string would prevent GC
+		// from collecting the whole large slice.
+		strings[i] = make([]byte, length)
+		_, err = r.Read(strings[i])
+		if err != nil {
+			break
+		}
+	}
+	switch i {
+	case 0: // `break` during prerelease string parsing
+		return ErrInvalidPrerelease
+	case 1: // `break` during metadata string parsing
+		return ErrInvalidMetadata
+	}
+
+	*v = Version{
+		major:    nums[0],
+		minor:    nums[1],
+		patch:    nums[2],
+		pre:      string(strings[0]),
+		metadata: string(strings[1]),
+	}
+	return nil
+}
+
+// MarshalBinary implements the encoding.BinaryMarshaler interface.
+func (v Version) MarshalBinary() ([]byte, error) {
+	// Once semver has 1.19 as a minimal supported go version -
```

Comment:
Or we can just copy [the obvious 5-line long function](https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/encoding/binary/varint.go;l=41) from golang (if the license permits that, of course 😇)